src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
+    {
+        private ReadOnlySequence<byte> _sequence;
+        private bool _isReaderCompleted;
+
+        private int _cancelNext;
+
+        public SequencePipeReader(ReadOnlySequence<byte> sequence)
+        {
+            _sequence = sequence;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
+
+            _sequence = _sequence.Slice(consumed);
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingRead()
+        {
+            Interlocked.Exchange(ref _cancelNext, 1);
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception? exception = null)
+        {
+            if (_isReaderCompleted)
+            {
+                return;
+            }
+
+            _isReaderCompleted = true;
+            _sequence = ReadOnlySequence<byte>.Empty;
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            if (TryReadInternal(out var result))
+            {
+                return new ValueTask<ReadResult>(result);
+            }
+
+            result = new ReadResult(_sequence, isCanceled: false, isCompleted: true);
```

Comment:
Nit for clarity:

```suggestion
            result = new ReadResult(ReadOnlySequence<byte>.Empty, isCanceled: false, isCompleted: true);
```

Or maybe a comment pointing out that `_sequence` is always empty here.

Reply:
I agree. Added the requested changes :)